File: adapters/unity/XaceConsoleWidget.cs
// XaceConsoleWidget.cs
// Lightweight in-game console for live XACE edit prompts.
//
// The widget owns only UI state. It emits explicit prompt/apply/cancel events
// and can optionally send CONTROL-style dictionaries through XaceTransport for
// future builder/runtime integrations.

using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using UnityEngine;

namespace Xace.Adapter.Unity
{
    public sealed class XaceConsoleWidget : MonoBehaviour
    {
        public enum ConsoleState
        {
            Idle,
            PromptSubmitted,
            PreviewReceived,
            UserDecision,
            Applying,
            Error
        }

        [Header("Input")]
        [SerializeField] private KeyCode toggleKey = KeyCode.F1;
        [SerializeField] private bool startVisible = false;
        [SerializeField] private bool sendControlToRuntime = false;

        [Header("Window")]
        [SerializeField] private Rect windowRect = new Rect(20f, 20f, 620f, 420f);
        [SerializeField] private int maxLogLines = 12;

        public event Action<string> OnPromptSubmitted;
        public event Action<string> OnApplyRequested;
        public event Action<string> OnCancelRequested;

        private XaceTransport transport;
        private ConsoleState state = ConsoleState.Idle;
        private bool visible;
        private string prompt = "";
        private string preview = "";
        private string mutationId = "";
        private string runtimeCgsHash = "";
        private string lastSnapshotHash = "";
        private ulong lastRuntimeTick;
        private float confidence;
        private string sessionId;
        private Vector2 logScroll;
        private readonly List<string> logLines = new List<string>();
        private readonly List<string> history = new List<string>();
        private int historyIndex = -1;
        private ulong controlSequence = 1;

        private GUIStyle labelStyle;
        private GUIStyle logStyle;
        private GUIStyle windowStyle;
        private bool stylesReady;

        public ConsoleState State => state;
        public bool IsVisible => visible;

        private void Awake()
        {
            transport = GetComponent<XaceTransport>() ?? FindAnyObjectByType<XaceTransport>();
            visible = startVisible;
            sessionId = Guid.NewGuid().ToString("N").Substring(0, 8);
        }

        private void OnEnable()
        {
            if (transport != null)
            {
                transport.OnHandshakeAccepted += OnHandshakeAccepted;
                transport.OnProtocolError += OnProtocolError;
                transport.OnMessageReceived += OnRuntimeMessage;
            }
        }

        private void OnDisable()
        {
            if (transport != null)
            {
                transport.OnHandshakeAccepted -= OnHandshakeAccepted;
                transport.OnProtocolError -= OnProtocolError;
                transport.OnMessageReceived -= OnRuntimeMessage;
            }
        }

        private void Update()
        {
            if (SafeKeyDown(toggleKey))
                visible = !visible;

            if (!visible)
                return;

            if (SafeKeyDown(KeyCode.UpArrow) && history.Count > 0)
            {
                historyIndex = Mathf.Clamp(historyIndex + 1, 0, history.Count - 1);
                prompt = history[history.Count - 1 - historyIndex];
            }
            else if (SafeKeyDown(KeyCode.DownArrow) && history.Count > 0)
            {
                historyIndex = Mathf.Clamp(historyIndex - 1, -1, history.Count - 1);
                prompt = historyIndex < 0 ? "" : history[history.Count - 1 - historyIndex];
            }
        }

        private void OnGUI()
        {
            if (!visible)
                return;
            EnsureStyles();
            windowRect = GUI.Window(915042, windowRect, DrawWindow, "XACE Console", windowStyle);
        }

        public void ReceivePreview(string mutationPreview, float previewConfidence, string previewMutationId = "")
        {
            preview = mutationPreview ?? "";
            confidence = Mathf.Clamp01(previewConfidence);
            mutationId = previewMutationId ?? "";
            state = ConsoleState.UserDecision;
            AppendLog("preview: " + preview);
        }

        public void SetError(string message)
        {
            state = ConsoleState.Error;
            AppendLog("error: " + message);
        }

        private void DrawWindow(int id)
        {
            GUILayout.Space(8f);
            GUILayout.Label("State: " + state + "  Runtime: " + ((transport != null && transport.IsConnected) ? "Connected" : "Disconnected"), labelStyle);
            GUILayout.Label("Tick: " + lastRuntimeTick + "  CGS: " + ShortHash(runtimeCgsHash) + "  Snapshot: " + ShortHash(lastSnapshotHash), labelStyle);
            GUILayout.Space(4f);

            GUI.enabled = state == ConsoleState.Idle || state == ConsoleState.Error;
            GUILayout.BeginHorizontal();
            GUILayout.Label("Prompt", GUILayout.Width(54f));
            GUI.SetNextControlName("XacePromptField");
            prompt = GUILayout.TextField(prompt ?? "", GUILayout.ExpandWidth(true));
            var submitPressed = GUILayout.Button("Send", GUILayout.Width(68f));
            GUILayout.EndHorizontal();
            GUI.enabled = true;

            if (submitPressed || (Event.current.type == EventType.KeyDown && Event.current.keyCode == KeyCode.Return))
            {
                if ((state == ConsoleState.Idle || state == ConsoleState.Error) && !string.IsNullOrWhiteSpace(prompt))
                    SubmitPrompt(prompt.Trim());
            }

            GUILayout.Space(8f);
            DrawConfidence();

            if (!string.IsNullOrEmpty(preview))
            {
                GUILayout.Label("Preview", labelStyle);
                GUILayout.TextArea(preview, GUILayout.MinHeight(72f));
                GUILayout.BeginHorizontal();
                GUI.enabled = state == ConsoleState.UserDecision;
                if (GUILayout.Button("Apply", GUILayout.Width(100f)))
                    Apply();
                if (GUILayout.Button("Cancel", GUILayout.Width(100f)))
                    Cancel();
                GUI.enabled = true;
                GUILayout.EndHorizontal();
            }

            GUILayout.Space(8f);
            GUILayout.Label("Log", labelStyle);
            logScroll = GUILayout.BeginScrollView(logScroll, GUILayout.ExpandHeight(true));
            foreach (var line in logLines)
                GUILayout.Label(line, logStyle);
            GUILayout.EndScrollView();

            if (GUI.Button(new Rect(windowRect.width - 30f, 2f, 24f, 20f), "x"))
                visible = false;
            GUI.DragWindow(new Rect(0f, 0f, windowRect.width - 34f, 24f));
        }

        private void DrawConfidence()
        {
            GUILayout.BeginHorizontal();
            GUILayout.Label("Confidence", GUILayout.Width(82f));
            var rect = GUILayoutUtility.GetRect(180f, 16f, GUILayout.Width(180f));
            var old = GUI.color;
            GUI.color = Color.gray;
            GUI.DrawTexture(rect, Texture2D.whiteTexture);
            GUI.color = confidence >= 0.7f ? Color.green : confidence >= 0.4f ? Color.yellow : Color.red;
            GUI.DrawTexture(new Rect(rect.x, rect.y, rect.width * confidence, rect.height), Texture2D.whiteTexture);
            GUI.color = old;
            GUILayout.Label((confidence * 100f).ToString("0") + "%");
            GUILayout.EndHorizontal();
        }

        private void SubmitPrompt(string text)
        {
            prompt = "";
            preview = "";
            confidence = 0f;
            mutationId = "";
            state = ConsoleState.PromptSubmitted;
            history.Add(text);
            if (history.Count > 50)
                history.RemoveAt(0);
            historyIndex = -1;
            AppendLog("> " + text);
            OnPromptSubmitted?.Invoke(text);
            SendControl("PromptSubmit", new SortedDictionary<string, object>(StringComparer.Ordinal)
            {
                ["prompt"] = text,
                ["session_id"] = sessionId
            });
        }

        private void Apply()
        {
            state = ConsoleState.Applying;
            AppendLog("apply requested");
            OnApplyRequested?.Invoke(mutationId);
            SendControl("MutationDecision", new SortedDictionary<string, object>(StringComparer.Ordinal)
            {
                ["decision"] = "Apply",
                ["mutation_id"] = mutationId,
                ["session_id"] = sessionId
            });
        }

        private void Cancel()
        {
            state = ConsoleState.Idle;
            AppendLog("cancelled");
            OnCancelRequested?.Invoke(mutationId);
            SendControl("MutationDecision", new SortedDictionary<string, object>(StringComparer.Ordinal)
            {
                ["decision"] = "Cancel",
                ["mutation_id"] = mutationId,
                ["session_id"] = sessionId
            });
            preview = "";
            mutationId = "";
            confidence = 0f;
        }

        private void SendControl(string controlType, SortedDictionary<string, object> payload)
        {
            if (!sendControlToRuntime || transport == null)
                return;
            payload["msg_type"] = "control";
            payload["control_type"] = controlType;
            payload["sequence_id"] = controlSequence++;
            transport.SendDictionary(payload);
        }

        private void OnHandshakeAccepted(XaceHandshakeAck ack)
        {
            runtimeCgsHash = ack.cgs_hash ?? "";
            AppendLog("connected: " + ack.session_id);
            AppendLog("cgs hash: " + ShortHash(runtimeCgsHash));
            if (state == ConsoleState.Error)
                state = ConsoleState.Idle;
        }

        private void OnProtocolError(string message)
        {
            SetError(message);
        }

        private void OnRuntimeMessage(XaceRuntimeMessage message)
        {
            if (message.MessageType == XaceProtocolNames.TickSnapshot)
            {
                lastRuntimeTick = message.Tick;
                lastSnapshotHash = SnapshotProofHash(message.RawJson);
                return;
            }

            if (message.MessageType == XaceProtocolNames.AdapterSideEffectRollback)
            {
                preview = "";
                mutationId = "";
                confidence = 0f;
                state = ConsoleState.Idle;
                AppendLog("rollback: " + message.GetString("reason", "adapter side effects restored"));
                return;
            }

            if (message.MessageType != "control")
                return;
            var controlType = message.GetString("control_type", "");
            if (controlType == "MutationPreview")
            {
                ReceivePreview(
                    message.GetString("description", message.GetString("message", "")),
                    Convert.ToSingle(message.Data.TryGetValue("confidence", out var c) ? c : 0f),
                    message.GetString("mutation_id", "")
                );
            }
            else if (controlType == "MutationApplied")
            {
                AppendLog("applied");
                state = ConsoleState.Idle;
                preview = "";
            }
            else if (controlType == "MutationCancelled")
            {
                AppendLog("cancelled by runtime");
                state = ConsoleState.Idle;
                preview = "";
            }
        }

        private void AppendLog(string line)
        {
            logLines.Add("[" + DateTime.Now.ToString("HH:mm:ss") + "] " + line);
            while (logLines.Count > maxLogLines)
                logLines.RemoveAt(0);
            logScroll = new Vector2(0f, float.MaxValue);
        }

        private static string SnapshotProofHash(string rawJson)
        {
            if (string.IsNullOrEmpty(rawJson))
                return "";
            using (var sha = SHA256.Create())
            {
                var bytes = sha.ComputeHash(Encoding.UTF8.GetBytes(rawJson));
                var builder = new StringBuilder(bytes.Length * 2);
                foreach (var b in bytes)
                    builder.Append(b.ToString("x2"));
                return builder.ToString();
            }
        }

        private static string ShortHash(string value)
        {
            return string.IsNullOrEmpty(value) ? "-" : value.Substring(0, Math.Min(12, value.Length));
        }

        private static bool SafeKeyDown(KeyCode key)
        {
            try { return Input.GetKeyDown(key); }
            catch (InvalidOperationException) { return false; }
            catch (ArgumentException) { return false; }
        }

        private void EnsureStyles()
        {
            if (stylesReady)
                return;
            windowStyle = new GUIStyle(GUI.skin.window)
            {
                fontSize = 13,
                padding = new RectOffset(10, 10, 22, 10)
            };
            labelStyle = new GUIStyle(GUI.skin.label)
            {
                fontStyle = FontStyle.Bold
            };
            logStyle = new GUIStyle(GUI.skin.label)
            {
                wordWrap = true,
                normal = { textColor = new Color(0.78f, 0.9f, 0.78f) }
            };
            stylesReady = true;
        }
    }
}

File: adapters/unreal/XaceConsoleWidget.cpp
#include "XaceConsoleWidget.h"

#include "Components/Button.h"
#include "Components/EditableTextBox.h"
#include "Components/ProgressBar.h"
#include "Components/ScrollBox.h"
#include "Components/TextBlock.h"
#include "Dom/JsonValue.h"
#include "Misc/SecureHash.h"
#include "Serialization/JsonSerializer.h"
#include "Serialization/JsonWriter.h"

void UXaceConsoleWidget::NativeOnInitialized()
{
	Super::NativeOnInitialized();
	SessionId = FGuid::NewGuid().ToString(EGuidFormats::Digits).Left(8);
	if (SubmitButton != nullptr)
	{
		SubmitButton->OnClicked.AddDynamic(this, &UXaceConsoleWidget::HandleSubmitClicked);
	}
	if (ApplyButton != nullptr)
	{
		ApplyButton->OnClicked.AddDynamic(this, &UXaceConsoleWidget::HandleApplyClicked);
	}
	if (CancelButton != nullptr)
	{
		CancelButton->OnClicked.AddDynamic(this, &UXaceConsoleWidget::HandleCancelClicked);
	}
	RefreshUi();
}

void UXaceConsoleWidget::NativeDestruct()
{
	if (Transport != nullptr)
	{
		Transport->OnHandshakeAccepted.RemoveDynamic(this, &UXaceConsoleWidget::HandleHandshake);
		Transport->OnProtocolError.RemoveDynamic(this, &UXaceConsoleWidget::HandleProtocolError);
		Transport->OnJsonMessage.RemoveAll(this);
	}
	Super::NativeDestruct();
}

void UXaceConsoleWidget::BindTransport(UXaceTransportComponent* InTransport)
{
	if (Transport == InTransport)
	{
		return;
	}
	if (Transport != nullptr)
	{
		Transport->OnHandshakeAccepted.RemoveDynamic(this, &UXaceConsoleWidget::HandleHandshake);
		Transport->OnProtocolError.RemoveDynamic(this, &UXaceConsoleWidget::HandleProtocolError);
		Transport->OnJsonMessage.RemoveAll(this);
	}
	Transport = InTransport;
	if (Transport != nullptr)
	{
		Transport->OnHandshakeAccepted.AddDynamic(this, &UXaceConsoleWidget::HandleHandshake);
		Transport->OnProtocolError.AddDynamic(this, &UXaceConsoleWidget::HandleProtocolError);
		Transport->OnJsonMessage.AddUObject(this, &UXaceConsoleWidget::HandleJsonMessage);
	}
}

void UXaceConsoleWidget::SubmitPrompt(const FString& Prompt)
{
	const FString Clean = Prompt.TrimStartAndEnd();
	if (Clean.IsEmpty())
	{
		return;
	}
	SetState(EXaceConsoleState::PromptSubmitted);
	PendingMutationId.Reset();
	CurrentConfidence = 0.0f;
	if (PreviewText != nullptr)
	{
		PreviewText->SetText(FText::GetEmpty());
	}
	AppendLog(TEXT("> ") + Clean);
	OnPromptSubmitted.Broadcast(Clean);

	TSharedRef<FJsonObject> Payload = MakeShared<FJsonObject>();
	Payload->SetStringField(TEXT("prompt"), Clean);
	Payload->SetStringField(TEXT("session_id"), SessionId);
	SendControl(TEXT("PromptSubmit"), Payload);
}

void UXaceConsoleWidget::ReceivePreview(const FString& Preview, float Confidence, const FString& MutationId)
{
	PendingMutationId = MutationId;
	CurrentConfidence = FMath::Clamp(Confidence, 0.0f, 1.0f);
	if (PreviewText != nullptr)
	{
		PreviewText->SetText(FText::FromString(Preview));
	}
	AppendLog(TEXT("preview: ") + Preview);
	SetState(EXaceConsoleState::UserDecision);
}

void UXaceConsoleWidget::SetConsoleError(const FString& Message)
{
	AppendLog(TEXT("error: ") + Message);
	SetState(EXaceConsoleState::Error);
}

void UXaceConsoleWidget::ApplyMutation()
{
	SetState(EXaceConsoleState::Applying);
	AppendLog(TEXT("apply requested"));
	OnApplyRequested.Broadcast(PendingMutationId);

	TSharedRef<FJsonObject> Payload = MakeShared<FJsonObject>();
	Payload->SetStringField(TEXT("decision"), TEXT("Apply"));
	Payload->SetStringField(TEXT("mutation_id"), PendingMutationId);
	Payload->SetStringField(TEXT("session_id"), SessionId);
	SendControl(TEXT("MutationDecision"), Payload);
}

void UXaceConsoleWidget::CancelMutation()
{
	AppendLog(TEXT("cancelled"));
	OnCancelRequested.Broadcast(PendingMutationId);

	TSharedRef<FJsonObject> Payload = MakeShared<FJsonObject>();
	Payload->SetStringField(TEXT("decision"), TEXT("Cancel"));
	Payload->SetStringField(TEXT("mutation_id"), PendingMutationId);
	Payload->SetStringField(TEXT("session_id"), SessionId);
	SendControl(TEXT("MutationDecision"), Payload);

	PendingMutationId.Reset();
	CurrentConfidence = 0.0f;
	SetState(EXaceConsoleState::Idle);
}

FString UXaceConsoleWidget::GetStateName() const
{
	switch (State)
	{
	case EXaceConsoleState::Idle: return TEXT("Idle");
	case EXaceConsoleState::PromptSubmitted: return TEXT("PromptSubmitted");
	case EXaceConsoleState::PreviewReceived: return TEXT("PreviewReceived");
	case EXaceConsoleState::UserDecision: return TEXT("UserDecision");
	case EXaceConsoleState::Applying: return TEXT("Applying");
	case EXaceConsoleState::Error: return TEXT("Error");
	default: return TEXT("Unknown");
	}
}

void UXaceConsoleWidget::HandleSubmitClicked()
{
	if (PromptInput == nullptr)
	{
		return;
	}
	const FString Prompt = PromptInput->GetText().ToString();
	PromptInput->SetText(FText::GetEmpty());
	SubmitPrompt(Prompt);
}

void UXaceConsoleWidget::HandleApplyClicked()
{
	ApplyMutation();
}

void UXaceConsoleWidget::HandleCancelClicked()
{
	CancelMutation();
}

void UXaceConsoleWidget::HandleHandshake(const FXaceHandshakeAck& Ack)
{
	RuntimeCgsHash = Ack.CgsHash;
	AppendLog(TEXT("connected: ") + Ack.SessionId);
	AppendLog(TEXT("cgs hash: ") + ShortHash(RuntimeCgsHash));
	if (State == EXaceConsoleState::Error)
	{
		SetState(EXaceConsoleState::Idle);
	}
}

void UXaceConsoleWidget::HandleProtocolError(const FString& Message)
{
	SetConsoleError(Message);
}

void UXaceConsoleWidget::HandleJsonMessage(const TSharedPtr<FJsonObject>& Message)
{
	FString MsgType;
	if (!Message.IsValid() || !Message->TryGetStringField(TEXT("msg_type"), MsgType))
	{
		return;
	}

	if (MsgType == TEXT("tick_snapshot"))
	{
		double TickValue = 0.0;
		Message->TryGetNumberField(TEXT("tick"), TickValue);
		LastRuntimeTick = int64(FMath::Max(0.0, TickValue));
		LastSnapshotHash = SnapshotProofHash(Message);
		RefreshUi();
		return;
	}

	if (MsgType == TEXT("adapter_side_effect_rollback"))
	{
		PendingMutationId.Reset();
		CurrentConfidence = 0.0f;
		if (PreviewText != nullptr)
		{
			PreviewText->SetText(FText::GetEmpty());
		}
		FString Reason;
		Message->TryGetStringField(TEXT("reason"), Reason);
		AppendLog(TEXT("rollback: ") + (Reason.IsEmpty() ? TEXT("adapter side effects restored") : Reason));
		SetState(EXaceConsoleState::Idle);
		return;
	}

	if (MsgType != TEXT("control"))
	{
		return;
	}

	FString ControlType;
	Message->TryGetStringField(TEXT("control_type"), ControlType);
	if (ControlType == TEXT("MutationPreview"))
	{
		FString Description;
		FString MutationId;
		double Confidence = 0.0;
		Message->TryGetStringField(TEXT("description"), Description);
		Message->TryGetStringField(TEXT("mutation_id"), MutationId);
		Message->TryGetNumberField(TEXT("confidence"), Confidence);
		ReceivePreview(
			Description,
			float(Confidence),
			MutationId
		);
	}
	else if (ControlType == TEXT("MutationApplied"))
	{
		AppendLog(TEXT("applied"));
		SetState(EXaceConsoleState::Idle);
	}
	else if (ControlType == TEXT("MutationCancelled"))
	{
		AppendLog(TEXT("cancelled by runtime"));
		SetState(EXaceConsoleState::Idle);
	}
}

void UXaceConsoleWidget::SetState(EXaceConsoleState NextState)
{
	State = NextState;
	RefreshUi();
}

void UXaceConsoleWidget::AppendLog(const FString& Line)
{
	LogLines.Add(FDateTime::Now().ToString(TEXT("[%H:%M:%S] ")) + Line);
	while (LogLines.Num() > MaxLogLines)
	{
		LogLines.RemoveAt(0);
	}
	if (LogScroll != nullptr)
	{
		LogScroll->ClearChildren();
		for (const FString& Item : LogLines)
		{
			UTextBlock* Text = NewObject<UTextBlock>(LogScroll);
			Text->SetText(FText::FromString(Item));
			LogScroll->AddChild(Text);
		}
		LogScroll->ScrollToEnd();
	}
}

void UXaceConsoleWidget::SendControl(const FString& ControlType, const TSharedRef<FJsonObject>& Payload)
{
	if (!bSendControlToRuntime || Transport == nullptr)
	{
		return;
	}
	Payload->SetStringField(TEXT("msg_type"), TEXT("control"));
	Payload->SetStringField(TEXT("control_type"), ControlType);
	Payload->SetNumberField(TEXT("sequence_id"), ControlSequence++);
	Transport->SendJsonObject(Payload);
}

void UXaceConsoleWidget::RefreshUi()
{
	if (StateText != nullptr)
	{
		const FString Proof = FString::Printf(
			TEXT("%s | Tick: %lld | CGS: %s | Snapshot: %s"),
			*GetStateName(),
			LastRuntimeTick,
			*ShortHash(RuntimeCgsHash),
			*ShortHash(LastSnapshotHash)
		);
		StateText->SetText(FText::FromString(Proof));
	}
	if (ConfidenceBar != nullptr)
	{
		ConfidenceBar->SetPercent(CurrentConfidence);
	}
	const bool bCanPrompt = State == EXaceConsoleState::Idle || State == EXaceConsoleState::Error;
	const bool bCanDecide = State == EXaceConsoleState::UserDecision;
	if (SubmitButton != nullptr)
	{
		SubmitButton->SetIsEnabled(bCanPrompt);
	}
	if (PromptInput != nullptr)
	{
		PromptInput->SetIsEnabled(bCanPrompt);
	}
	if (ApplyButton != nullptr)
	{
		ApplyButton->SetIsEnabled(bCanDecide);
	}
	if (CancelButton != nullptr)
	{
		CancelButton->SetIsEnabled(bCanDecide || State == EXaceConsoleState::PromptSubmitted);
	}
}

FString UXaceConsoleWidget::SnapshotProofHash(const TSharedPtr<FJsonObject>& Message)
{
	if (!Message.IsValid())
	{
		return FString();
	}
	const FString Raw = JsonToString(Message.ToSharedRef());
	FMD5 Md5;
	FTCHARToUTF8 Bytes(*Raw);
	Md5.Update(reinterpret_cast<const uint8*>(Bytes.Get()), Bytes.Length());
	uint8 Digest[16];
	Md5.Final(Digest);

	FString Out;
	for (uint8 Byte : Digest)
	{
		Out += FString::Printf(TEXT("%02x"), Byte);
	}
	return Out;
}

FString UXaceConsoleWidget::ShortHash(const FString& Value)
{
	return Value.IsEmpty() ? TEXT("-") : Value.Left(12);
}

FString UXaceConsoleWidget::JsonToString(const TSharedRef<FJsonObject>& Object)
{
	FString Out;
	TSharedRef<TJsonWriter<>> Writer = TJsonWriterFactory<>::Create(&Out);
	FJsonSerializer::Serialize(Object, Writer);
	return Out;
}
